presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,225 @@
+import com.teradata.tempto.query.QueryExecutionException;
+import io.airlift.http.client.HttpStatus;
+import io.airlift.http.client.Response;
+import io.airlift.log.Logger;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.tests.TestGroups.CANCEL_QUERY;
+import static com.teradata.tempto.assertions.QueryAssert.assertThat;
+import static com.teradata.tempto.query.QueryExecutor.query;
+import static java.lang.String.format;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class SqlCancelTests
+        extends ProductTest
+{
+    private static String tableName = "table_name";
+
+    private static ExecutorService executor;
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    @AfterTestWithContext
+    public void dropTestTables()
+    {
+        try {
+            query(format("DROP TABLE IF EXISTS %s", tableName));
+        }
+        catch (Exception e) {
+            Logger.get(getClass()).warn(e, "failed to drop table");
+        }
+    }
+
+    @BeforeTestWithContext
+    public void startExecutor()
+    {
+        executor = Executors.newFixedThreadPool(2);
+    }
+
+    @AfterTestWithContext
+    public void stopExecutor()
+    {
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+            }
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("failed to stop the executor", e);
+        }
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Boolean>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException ExceptionDueToQueryCancellation) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+        private long waitTime;
+        private long step;
+
+        public SqlCancelCallable(String sqlStatement, long waitTime, long step)
+        {
+            this.sqlStatement = sqlStatement;
+            this.waitTime = waitTime;
+            this.step = step;
+        }
+
+        public Response call()
+        {
+            long time = 0;
+            while (time < waitTime) {
```

Comment:
This is not quite correct, since it does not account for the amount of time spent processing (just the time spent sleeping). It's better to do it based on the current system time. It's even better to use a library to do retries like this for you.